Source/Core/InputCommon/ControllerInterface/Win32/Win32.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "InputCommon/ControllerInterface/Win32/Win32.h"
+
+#include <windows.h>
+
+#include <thread>
+
+#include "Common/Flag.h"
+#include "Common/Logging/Log.h"
+#include "Common/ScopeGuard.h"
+#include "InputCommon/ControllerInterface/DInput/DInput.h"
+#include "InputCommon/ControllerInterface/XInput/XInput.h"
+
+constexpr UINT WM_DOLPHIN_STOP = WM_USER;
+
+static HWND s_hwnd;
+static HWND s_message_window;
+static std::thread s_thread;
+
+static LRESULT CALLBACK WindowProc(HWND hwnd, UINT message, WPARAM wparam, LPARAM lparam)
+{
+  if (message == WM_INPUT_DEVICE_CHANGE)
+  {
+    ciface::DInput::PopulateDevices(s_hwnd);
+    ciface::XInput::PopulateDevices();
+    g_controller_interface.InvokeHotplugCallbacks();
+  }
+
+  return DefWindowProc(hwnd, message, wparam, lparam);
+}
+
+void ciface::Win32::Init(void* hwnd)
+{
+  s_hwnd = static_cast<HWND>(hwnd);
+  XInput::Init();
+
+  s_thread = std::thread([] {
+    if (FAILED(CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED)))
+    {
+      ERROR_LOG(SERIALINTERFACE, "CoInitializeEx failed: %i", GetLastError());
+      return;
+    }
+    Common::ScopeGuard uninit([] { CoUninitialize(); });
+
+    WNDCLASSEX window_class_info;
```

Comment:
Maybe consider making this `WNDCLASSEX window_class_info{};`, so the `= 0` and `= nullptr` assignments can be removed? Feels a bit too verbose right now.

Reply:
Yeah, good call